s3transfer/tasks.py

```diff
@@ -279,6 +284,20 @@ def _submit(self, transfer_future, **kwargs):
         """
         raise NotImplementedError('_submit() must be implemented')
 
+    def _conditionally_execute_main(self, kwargs):
+        # If the task is not done (really only if some other related
+        # task to the TransferFuture had failed) then execute the task's
+        # main() method.
+        if not self._transfer_coordinator.done():
+            return self._execute_main(kwargs)
+        # If the transfer coordinator signals it is done prior to the
+        # submission task running _main, the submission task will not
+        # submit any further tasks and therefore must take responsibility
+        # to announce done because there will be no tasks submitted to
+        # announce done for the transfer.
+        else:
+            self._transfer_coordinator.announce_done()
```

Comment:
This seems odd to me.  We're saying if the transfer coordinator is not done, then run `_execute_main`.  However if the transfer coordinator *is* done, then `announce_done`.  Seems like the coordinator shouldn't be allowed to be in a terminal state without something calling `announce_done` (or it's equivalent).